From daff32411b71d7f9a0b4f7891c08011d690bd16f Mon Sep 17 00:00:00 2001
From: Thinkofdeath <thinkofdeath@spigotmc.org>
Date: Sun, 29 Jun 2014 21:10:34 +0100
Subject: [PATCH] Limit block placement/interaction packets


diff --git a/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java b/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
index cc4b6ef05..53a40cfcc 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
@@ -9,7 +9,13 @@ import net.minecraft.network.protocol.PacketType;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.phys.BlockHitResult;
 
-public record ServerboundUseItemOnPacket(InteractionHand hand, BlockHitResult hitResult, int sequence) implements Packet<ServerGamePacketListener> {
+// Spigot start
+public record ServerboundUseItemOnPacket(InteractionHand hand, BlockHitResult hitResult, int sequence, long timestamp) implements Packet<ServerGamePacketListener> {
+
+    public ServerboundUseItemOnPacket(InteractionHand hand, BlockHitResult hitResult, int sequence) {
+        this(hand, hitResult, sequence, System.currentTimeMillis());
+    }
+    // Spigot end
 
     public static final StreamCodec<FriendlyByteBuf, ServerboundUseItemOnPacket> STREAM_CODEC = StreamCodec.composite(InteractionHand.STREAM_CODEC, ServerboundUseItemOnPacket::hand, BlockHitResult.STREAM_CODEC, ServerboundUseItemOnPacket::hitResult, ByteBufCodecs.VAR_INT, ServerboundUseItemOnPacket::sequence, ServerboundUseItemOnPacket::new);
 
diff --git a/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java b/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
index caab94c53..5b59e94ef 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -8,7 +8,13 @@ import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 import net.minecraft.world.InteractionHand;
 
-public record ServerboundUseItemPacket(InteractionHand hand, int sequence, float yRot, float xRot) implements Packet<ServerGamePacketListener> {
+// Spigot start
+public record ServerboundUseItemPacket(InteractionHand hand, int sequence, float yRot, float xRot, long timestamp) implements Packet<ServerGamePacketListener> {
+
+    public ServerboundUseItemPacket(InteractionHand hand, int sequence, float yRot, float xRot) {
+        this(hand, sequence, yRot, xRot, System.currentTimeMillis());
+    }
+    // Spigot end
 
     public static final StreamCodec<FriendlyByteBuf, ServerboundUseItemPacket> STREAM_CODEC = StreamCodec.composite(InteractionHand.STREAM_CODEC, ServerboundUseItemPacket::hand, ByteBufCodecs.VAR_INT, ServerboundUseItemPacket::sequence, ByteBufCodecs.FLOAT, ServerboundUseItemPacket::yRot, ByteBufCodecs.FLOAT, ServerboundUseItemPacket::xRot, ServerboundUseItemPacket::new);
 
diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 675d2052a..ad9a68e4a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1916,10 +1916,30 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
         }
     }
 
+    // Spigot start - limit place/interactions
+    private int limitedPackets;
+    private long lastLimitedPacket = -1;
+
+    private boolean checkLimit(long timestamp) {
+        if (lastLimitedPacket != -1 && timestamp - lastLimitedPacket < 30 && limitedPackets++ >= 4) {
+            return false;
+        }
+
+        if (lastLimitedPacket == -1 || timestamp - lastLimitedPacket >= 30) {
+            lastLimitedPacket = timestamp;
+            limitedPackets = 0;
+            return true;
+        }
+
+        return true;
+    }
+    // Spigot end
+
     @Override
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.player.isImmobile()) return; // CraftBukkit
+        if (!checkLimit(packet.timestamp())) return; // Spigot - check limit
         if (this.hasClientLoaded()) {
             this.ackBlockChangesUpTo(packet.sequence());
             ServerLevel serverlevel = this.player.level();
@@ -2001,6 +2021,7 @@ public class ServerGamePacketListenerImpl extends ServerCommonPacketListenerImpl
     public void handleUseItem(ServerboundUseItemPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.player.isImmobile()) return; // CraftBukkit
+        if (!checkLimit(packet.timestamp())) return; // Spigot - check limit
         if (this.hasClientLoaded()) {
             this.ackBlockChangesUpTo(packet.sequence());
             ServerLevel serverlevel = this.player.level();
-- 
2.55.0

